### PR TITLE
API ideas

### PR DESCRIPTION
**Pull request by meejah**
(Note: this PR is intended to be closed; here for diff-viewing purposes).

After some brief feedback on #twisted, I've tried to simplify the API a little (some suggested it was "too wordy"). There was also suggestions to provide a non-magic, simple wrapper around the Future or Deferred (e.g. so you can call `.add_callbacks` on it).

Note I collapsed `create_future, create_future_success and create_future_error` down to a single API with a couple kwargs -- maybe it's better to keep them separate...

I also took a stab at a Mixin-style API. I prefixed everything with "promise_", which hopefully means it's conflict-free for most user-classes. This would be most-like the current Autobahn API...

I'm very strongly leaning towards calling everything a "promise" in the API/docs to lessen the chance of confusion between Future and Deferred, and to make it more-obvious this is its own API and not a "mere" compatilibity-wrapper (e.g. giving you Twisted-style API on top of asyncio or vice-versa).

@oberstet or @hawkowl what do you think?

### CODE REVIEW

**Comment by meejah** [comment]
I'm not really sure if I like the "wrap each Deferred or Future" idea, but I guess it would at least provide a consistent API. One _could_ attach an "add_callbacks" method to the Future or Deferred directly (in create_promise), I guess, but that's kind of weird too.

**Comment by oberstet** [comment]
> I'm not really sure if I like the "wrap each Deferred or Future" idea

No, please not. Wrapping a Twisted Deferred and asyncio Future in another class .. no. We won't be able to paper over the subtle differences anyway (I think). `txaio.tx.create_future` should return a plain Twisted Deferred, whereas `txaio.aio.create_future` should return a plain asyncio Future.

Regarding introducing yet another "name" with "promise": I am also not convinced this is a good idea. Eg. ES6 distinguishes between deferreds, which can be resolved/rejected and _hold_ a promise, which is where you attach handlers to. A Twisted Deferred plays both of the former roles. Different from asyncio Future's, Twisted Deferreds also have that funky callback/errback chains. It's all a big mess anyway;)

**Comment by meejah** [comment]
Yeah.

I guess that the point of the "wrap" thing was that you'd only have one or two methods to call (e.g. add_callbacks, .future, .deferred) and could still get to the "original" one -- so it'd be more obvious when you were doing something event-loop-specific. The harder difference to get around is the chaining... 

Speaking of the callback chaining: I have code that (sort of) handles that  consistently (by making Futures behave like Twisted, basically). However, it just does `future._result = newresult` essentially, which is obviously messing with internal state. But errbacks would be harder, and anyway I can't see any way to "cancel" an error in asyncio at all (well, I guess you could twiddle more internal state, but ...)

I was thinking about this, and it's kind-of tricky: if the point is to allow one to write consistent code betwen Twisted/asyncio then you either need to know what's going to happen with the callback results (i.e. either always chaining or alwaynot), or you simply shouldn't ever be allowed to add _two_ callbacks to a Future/Deferred...which then might point towards having a wrapper to enforce this :/

For now I guess the best course of action is just documenting it (?) since Autobahn doesn't rely on chaining callbacks anyway...but if it ever did, you'd get different behavior on asyncio and Twisted...

**Comment by oberstet** [comment]
> you either need to know what's going to happen with the callback results

Yes. I think trying to unify Twisted/asyncio retrospect / in a library would lead to insanity;) Guido has done a long research into Twisted Deferreds and in the end came up with something "different". I don't feel like questioning this - I take it as a fact.

> For now I guess the best course of action is just documenting it 

Yes.

> Autobahn doesn't rely on chaining callbacks

Yes. Deliberately.

**Comment by meejah** [comment]
Okay, I'll just document that you should probably just "return arg" in all your Twisted callbacks (if you use multiple ones) to get consistent asyncio/Twisted behavior.

...as for naming, i can change any mention of "promise" back to "future" in the API. Hopefully not confusing ;)

As for the "wordiness" comments, etc. and the Mixins: I tried to make the ILoopMixin consistently start with "promise_" (which I'll make "future_"). Should I proceed with that, and use them in Autobahn? Or is the "self.txaio" way (i.e. get rid of the mixins) the one to go with?

How about shortening the `txaio` APIs to remove "future" since it's theoretically obvious what you're doing. That is "txaio.add_callbacks", "txaio.reject" etc. instead of "add_future_callbacks", "reject_future" etc.

**Comment by oberstet** [comment]
Just have read your proposal

``` python
    from txaio import tx as _txaio
    class ApplicationSession(...):
         txaio = _txaio
```

If we do that, then I guess we don't need mixins. I like it. However, don't we want "call_later" also? That needs an reactor/loop. See these two badly misplaced pieces

https://github.com/tavendo/AutobahnPython/blob/master/autobahn/asyncio/websocket.py#L219
https://github.com/tavendo/AutobahnPython/blob/master/autobahn/twisted/websocket.py#L196

If we need reactor/loop, shouldn't the user be able to provide that?

https://github.com/tavendo/AutobahnPython/pull/363#issuecomment-82955593

Mmh. I am still not sure what's the "best" approach. Your proposal above is simple and nice, and does away with mixins. But the call_later thing (at least) .. those misplaced code pieces above should also go somewhere else.

Misc:
- naming should probably follow pep8 .. no more CamelCase for methods
- pls absolute imports everywhere .. https://github.com/meejah/txaio/blob/master/txaio/aio.py#L7
- https://github.com/meejah/txaio/blob/master/txaio/tx.py#L96 why all this? Similar to create_future, gather_future should return a native Twisted DeferredList .. and the user has to be ready to cope with that.

**Comment by meejah** [comment]
Regarding the gather() handling: overall, I was trying to make `txaio` provide a consistent API -- no matter if you're using asyncio or Twisted "under the hood". Ditto with callback chaining. Otherwise, I don't know if I see the point of `txaio`?

This was also why I strongly lean towards using "promise" in the API: so that it's specifically _not_ familiar to either asyncio or Twisted users and, hopefully, they don't expect it to behave like "the thing they're familiar with". Because it won't ;)

So anyway for gather: Twisted will return, by default, a list of 2-tuples: (ok, value) where ok is True or False, and value is the value or the Failure. Whereas asyncio doesn't do that. So one or the other implementation needs some code (or else, you have to change the callback you add to the promise/future returned from gather() depending on whether you're using asyncio or Twisted -- and again, then I don't see the point of txaio in such a case).

As far as "call_later". Hmm. Yes. It will need an event-loop, unfortunately. 

I could get rid of the first_exception and/or first_error stuff -- I'm using it in CTS for timeouts (but gather() could take a timeout instead -- as asyncio supports that directly, whereas Twisted does not support a timeout to DeferredList).

Oh, naming: IFailedFuture is following Twisted Failure because it actually _is_ just a failure in that case. Maybe this really does point to needing wrappers? It's still possible to do "print_traceback" instead, but then I'd need to either wrap Twisted Failures or attach a method to them (e.g. in create_failure()). Alternatively, we could just use FailedPromise as the arg to errback() always ... and provide an implementation for txaio.tx too... Hmmm.

I'll come up with a more-comprehensive overall thing tonight. I completely missed the call_later() API obviously, so I'll have to deal with that as well.
